staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go

```diff
@@ -222,7 +222,13 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		// choose to not enable OpenAPI by having null openAPIConfig, and thus OpenAPIVersionedService
 		// and StaticOpenAPISpec are both null. In that case we don't run the CRD OpenAPI controller.
 		if s.GenericAPIServer.OpenAPIVersionedService != nil && s.GenericAPIServer.StaticOpenAPISpec != nil {
-			go openapiController.Run(s.GenericAPIServer.StaticOpenAPISpec, s.GenericAPIServer.OpenAPIVersionedService, context.StopCh)
+			// Wait for first sync to finish in order to serve a complete OpenAPI spec before ready
```

Comment:
Can I ask why only the CRD openapi controller is blocked on? Even if the openapi for CRDs is ready, it doesn't guarantee that the aggregator will have the aggregated openapi ready at the `openapi/v2` endpoint? (I might be misunderstanding something here since I remember your original approach directly modified the aggregator)

Reply:
@Jefftree, after discussing with @sttts, as you suspect there is still a race. Plausible option is to make this initial aggregation synchronous(from apiextension -> aggregator). Otherwise, there will always be race.